osquery/tables/applications/browser_utils.cpp

```diff
@@ -172,7 +172,58 @@ void genExtension(const std::string& uid,
 }
 
 QueryData genChromeBasedExtensions(QueryContext& context,
-                                   const fs::path& sub_dir) {
+                                   const std::vector<fs::path>& chromePaths) {
+  QueryData results;
+
+  auto users = usersFromContext(context);
+  for (const auto& row : users) {
+    if (row.count("uid") > 0 && row.count("directory") > 0) {
+      // For each user, enumerate all of their chrome profiles.
+      std::vector<std::string> profiles;
+      for (const auto& chromePath : chromePaths) {
+        fs::path extension_path = row.at("directory") / chromePath;
+        if (!resolveFilePattern(extension_path, profiles, GLOB_FOLDERS).ok()) {
+          continue;
+        }
+
+        // For each profile list each extension in the Extensions directory.
+        for (const auto& profile : profiles) {
+          std::vector<std::string> extensions = {};
+          listDirectoriesInDirectory(profile, extensions);
+
+          if (extensions.empty()) {
+            continue;
+          }
+
+          auto profile_path = fs::path(profile).parent_path().parent_path();
+
+          std::string profile_name;
+          auto status = getChromeProfileName(profile_name, profile_path);
+          if (!status.ok()) {
+            LOG(WARNING) << "Getting Chrome profile name failed: "
+                         << status.getMessage();
+          }
+
+          // Generate an addons list from their extensions JSON.
+          std::vector<std::string> versions;
+          for (const auto& extension : extensions) {
+            listDirectoriesInDirectory(extension, versions);
+          }
+
+          // Extensions use /<EXTENSION>/<VERSION>/manifest.json.
+          for (const auto& version : versions) {
+            genExtension(row.at("uid"), version, profile_name, results);
+          }
+        }
+      }
+    }
+  }
+
+  return results;
+}
+
+QueryData genOperaBasedExtensions(QueryContext& context,
```

Comment:
I recommend a different approach here that results in less code duplication. I'll leave _First_, _Second_, and _Third_ as inline comments below to describe my suggestion.

(First) delete this method `genOperaBasedExtensions` (lines 225-272).